File: backend/lib/tools/update-form-field.ts
import { getSql } from "../db";
import { applyFieldPath, validateFieldPath } from "../forms/field-path";
import type { FormType } from "@contract/contract";
import type { ToolEmit } from "./index";

type Input = {
  ticket_id: number;
  form_type: FormType;
  field_path: string;
  value: unknown;
  source_message_id: number;
};

export async function updateFormField(inp: Input, emit: ToolEmit) {
  if (!validateFieldPath(inp.form_type, inp.field_path)) {
    return { error: `invalid field_path '${inp.field_path}' for form ${inp.form_type}` };
  }
  const sql = getSql();
  const rows = await sql<{ payload: any }[]>`
    SELECT payload FROM forms WHERE ticket_id = ${inp.ticket_id} AND form_type = ${inp.form_type}
  `;
  if (!rows[0]) return { error: `form ${inp.form_type} not found for ticket ${inp.ticket_id}` };

  const payload = rows[0].payload; // jsonb returns parsed object
  const updated = applyFieldPath(payload, inp.field_path, inp.value);
  const now = new Date().toISOString();
  await sql`
    UPDATE forms
    SET payload = ${sql.json(updated as any)}, updated_at = ${now}
    WHERE ticket_id = ${inp.ticket_id} AND form_type = ${inp.form_type}
  `;

  emit({
    type: "form_updated",
    form_type: inp.form_type,
    payload: updated as Record<string, unknown>,
  });

  return { ok: true, form_type: inp.form_type, field_path: inp.field_path };
}

File: backend/railio/tools/update_form_field.py
"""Update one field on a form."""

from __future__ import annotations

from typing import Any, Callable

from sqlalchemy import bindparam, text
from sqlalchemy.dialects.postgresql import JSONB

from ..contract import FormType
from ..db import session_scope
from ..field_path import apply_field_path, validate_field_path
from ..messages_repo import _iso_now

ToolEmit = Callable[[dict[str, Any]], None]


async def update_form_field(
    *,
    ticket_id: int,
    form_type: FormType,
    field_path: str,
    value: Any,
    source_message_id: int,
    emit: ToolEmit,
) -> dict[str, Any]:
    if not validate_field_path(form_type, field_path):
        return {"error": f"invalid field_path '{field_path}' for form {form_type}"}

    async with session_scope() as session:
        row = (
            await session.execute(
                text(
                    """
                    SELECT payload FROM forms
                    WHERE ticket_id = :tid AND form_type = :ft
                    """
                ),
                {"tid": ticket_id, "ft": form_type},
            )
        ).mappings().first()
        if not row:
            return {"error": f"form {form_type} not found for ticket {ticket_id}"}

        payload = row["payload"] if isinstance(row["payload"], dict) else dict(row["payload"])
        updated = apply_field_path(payload, field_path, value)
        now = _iso_now()
        stmt = (
            text(
                """
                UPDATE forms SET payload = :payload, updated_at = :now
                WHERE ticket_id = :tid AND form_type = :ft
                """
            )
            .bindparams(bindparam("payload", type_=JSONB))
        )
        await session.execute(
            stmt, {"payload": updated, "now": now, "tid": ticket_id, "ft": form_type}
        )

    emit({"type": "form_updated", "form_type": form_type, "payload": updated})
    return {"ok": True, "form_type": form_type, "field_path": field_path}
